File: LeadToOffer/20_MinInStack.py
## min array always keep the min up to now.
class Solution:
    def __init__(self):
        self.vals = []
        self.mins = []

    def push(self, node):
        self.vals.append(node)
        if len(self.mins)==0 or node<self.mins[-1]:
            self.mins.append(node)
        else:
            self.mins.append(self.mins[-1])

    def pop(self):
        self.vals.pop()
        self.mins.pop()

    def top(self):
        return self.vals[-1]

    def min(self):
        return self.mins[-1]

File: Algo/LeadToOffer/20_MinInStack.cpp
// min array always keep the min up to now.
class Solution {
public:
    void push(int value) {
        vals.push(value);
        if (mins.empty() || mins.top()>value)
            mins.push(value);
        else
            mins.push(mins.top());
    }
    void pop() {
        vals.pop();
        mins.pop();
    }
    int top() {
        return vals.top();
    }
    int min() {
        return mins.top();
    }
private:
    stack<int> vals;
    stack<int> mins;
};

// 2. You do not need to push all the same mins.
class Solution {
public:
    void push(int value) {
        vals.push(value);
        if (mins.empty() || mins.top()>=value)
            mins.push(value);
    }
    void pop() {
        if (vals.top() == mins.top())
            mins.pop();
        vals.pop();
    }
    int top() {
        return vals.top();
    }
    int min() {
        return mins.top();
    }
private:
    stack<int> vals;
    stack<int> mins;
};
